cmd/deploy.go

```diff
@@ -354,6 +356,68 @@ func autofillApimodel(dc *deployCmd) error {
 			}
 		}
 	}
+
+	if k8sConfig != nil && k8sConfig.Addons != nil && k8sConfig.IsContainerMonitoringAddonEnabled() {
+		log.Infoln("container monitoring addon enabled")
+		addon := k8sConfig.GetAddonByName("container-monitoring")
+		if addon.Config == nil || len(addon.Config) == 0 || addon.Config["logAnalyticsWorkspaceResourceId"] != "" {
+			workspaceResourceId := strings.TrimSpace(addon.Config["logAnalyticsWorkspaceResourceId"])
+			if workspaceResourceId != "" {
+				log.Infoln("using provided log analytics workspace resource id:", workspaceResourceId)
+				if !strings.HasPrefix(workspaceResourceId, "/") {
+					workspaceResourceId = "/" + workspaceResourceId
+				}
+				workspaceResourceId = strings.TrimSuffix(workspaceResourceId, "/")
+			} else {
+				log.Infoln("creating default log analytics workspace")
+				workspaceResourceId, err = dc.client.EnsureDefaultLogAnalyticsWorkspace(ctx, dc.resourceGroup, dc.location)
+				if err != nil {
+					return err
+				}
+				log.Infoln("successfully created default log analytics workspace:", workspaceResourceId)
+			}
+
+			resourceParts := strings.Split(workspaceResourceId, "/")
+			if len(resourceParts) != 9 {
+				return fmt.Errorf("%s is not a valid resourceID", workspaceResourceId)
+			}
+			workspaceSubscriptionId := resourceParts[2]
+			workspaceResourceGroup := resourceParts[4]
+			workspaceName := resourceParts[8]
+			log.Infoln("Retrieving log analytics workspace Guid, Key and location details")
+			wsId, wsKey, wsLocation, err := dc.client.GetLogAnalyticsWorkspaceInfo(ctx, workspaceSubscriptionId, workspaceResourceGroup, workspaceName)
```

Comment:
deploy.go seems like the wrong place to do this as it's the CLI implementation for aks-engine. Some aks-engine users (such as AKS) don't use the CLI and instead call generate + az deployment create.
Also, this is creating Azure resources as far as I understand, it doesn't belong in the `autofillApimodel` function which is simply merging the apimodel with the command line flags provided by the user.

Reply:
This change primarily for the AKS-Engine customers and not for AKS.  Followed the same pattern as Service Principal, role assignment and RG resource creation logic here. This piece of the code activated only if the user has the Container Monitoring add-on configuration in the API definition file.  For the monitoring add-on, we would need to fetch the Key of Log analytics Workspace  i.e. user provided or created by default.  Based on my understanding, this is not possible to make the  ARM APIs calls to during the generate experience.

Please let me know if you have any concerns or suggestions on this implementation.

Reply:
Oh I see now. I guess this function is badly named (that's our fault) and is doing a lot more than it should. I'm still worried that this functionality relies entirely on using aks-engine deploy which not all customers do (not just AKS but also aks-engine customers). If that's the case, we should make it clear in the docs.

@mboersma thoughts?